go/cmd/launcher/main.go

```diff
@@ -0,0 +1,102 @@
+// Copyright (C) 2019 Red Hat, Inc.
+// SPDX-License-Identifier: LGPL-2.1-or-later
+
+package main
+
+import (
+	"errors"
+	"flag"
+	"fmt"
+	"github.com/virtblocks/virtblocks/go/pkg/command"
+	"github.com/virtblocks/virtblocks/go/pkg/devices"
+	"github.com/virtblocks/virtblocks/go/pkg/vm"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+)
+
+func processFlags() (disk string, cpus uint, mem uint, err error) {
+	flag.StringVar(&disk, "disk", "", "path to disk image (qcow2)")
+	flag.UintVar(&cpus, "cpus", 1, "number of CPUs")
+	flag.UintVar(&mem, "mem", 128, "amount of memory (MiB)")
+
+	flag.Parse()
+
+	if disk == "" {
+		err = errors.New("need a disk")
+		return
+	}
+	if filepath.Ext(disk) != ".qcow2" {
+		err = errors.New("need disk in qcow2 format")
+		return
+	}
+	if cpus < 1 {
+		err = errors.New("need at least one CPU")
+		return
+	}
+	if mem < 64 {
+		err = errors.New("need more memory")
+		return
+	}
+
+	return
+}
+
+func main() {
+	diskPath, cpus, mem, err := processFlags()
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "error processing flags: %v\n", err)
+		os.Exit(1)
+	}
+
+	// Derive the path of the socket used for serial console communication
+	// from the path of the disk image
+	serialSockPath := fmt.Sprintf("%s%s", strings.TrimSuffix(diskPath, "qcow2"), "serial-sock")
+
+	vm := vm.NewDescription()
+	vm.SetCpus(cpus).SetMemory(mem)
+	vm.SetDisk(devices.NewDisk().SetFilename(diskPath))
+	vm.SetSerial(devices.NewSerial().SetPath(serialSockPath))
+
+	qemuArgs, err := vm.QemuCommandLine()
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "error generating QEMU command: %v\n", err)
+		os.Exit(1)
+	}
+
+	qemu := command.NewCommand(qemuArgs[0])
+	qemu.AddArgs(qemuArgs[1:]...)
+
+	socat := command.NewCommand("socat")
```

Comment:
Would be nice to directly use golang here. Should work with something like this:

```go
	c, err := net.Dial("unix", "/my.sock")
	if err != nil {
		log.Fatalf("failed to connect to the the socket: %v", err)
	}
	defer c.Close()

	done := make(chan struct{})
	buf := make([]byte, 1024)
	go func(r io.Reader, done chan struct{}) {
		for {
			n, err := r.Read(buf[:])
			if err != nil {
				if err != io.EOF {
					log.Fatalf("failed to read from socket: %v", err)
				} else {
					done <- struct{}{}
				}
				return
			}
			fmt.Print(string(buf[0:n]))
		}
	}(c, done)

	writer := bufio.NewWriter(c)
	_, err = writer.ReadFrom(os.Stdin)
	if err != nil {
		log.Fatalf("failed to write to the socket: %v", err)
	}
	err = writer.Flush()
	if err != nil {
		log.Fatalf("failed to flush: %v", err)
	}
	// send EOF (I guess?), or do I you have to wait for it when reading from the socket?
	_, err = c.Write([]byte{0x11})
	if err != nil {
		log.Fatalf("failed to send EOF: %v", err)
	}
	// wait until we have read and printed everything
	<-done
```

Reply:
This is... Quite a lot, at least for someone who's not very familar with Go yet :)

Let's keep in mind that this launcher is just a silly demo, not something that we expect to be used for anything serious. So if we were to build this kind of functionality into out APIs, we'd definitely not be spawning an external command, but in the context of the silly demo I think the three-lines version is quite fine.

How about I add a comment on top clarifying that you're not supposed to do things this way in serious contexts? I already have a couple such comments in the program.